mysite/psa/pipeline.py

```diff
@@ -0,0 +1,296 @@
+            # This is very straightforward method
+            # TODO Need to check current user to avoid unnecessary check
+            if code.user_id:
+                user_from_code = User.objects.filter(id=code.user_id).first()
+                if user_from_code:
+                    user = user_from_code
+                    logout(backend.strategy.request)
+                    user.backend = 'django.contrib.auth.backends.ModelBackend'
+                    login(backend.strategy.request, user)
+                    return {'user': user}
+        else:
+            if user and user.groups.filter(name='Temporary').exists():
+                AnonymEmail.objects.get_or_create(user=user, email=details.get('email'),
+                                                  defaults={'date': datetime.now()})
+            backend.strategy.send_email_validation(backend, details['email'])
+            backend.strategy.session_set('email_validation_address',
+                                         details['email'])
+            return backend.strategy.redirect(
+                backend.strategy.setting('EMAIL_VALIDATION_URL')
+            )
+
+
+def union_merge(tmp_user, user):
+    """Union merge
+
+    Merging Roles, UnitStatus, FSMState, Response, StudentError objects.
+    """
+
+    """
+       Reassigning Roles
+       Doing UNION merge to not repeat roles to the same course
+       with the save role
+    """
+    roles_to_reset = (role for role in tmp_user.role_set.all()
+                      if not user.role_set.filter(course=role.course,
+                                                  role=role.role))
+    for role in roles_to_reset:
+        role.user = user
+        role.save()
+    """
+       Reassigning UnitStatuses
+       TODO think about filter() for UNION instead all()
+       maybe we need fresh unitstatuses instead of all to reassign
+    """
+    unitstatus_to_reset = (us for us in tmp_user.unitstatus_set.all())
+    for ut in unitstatus_to_reset:
+        ut.user = user
+        ut.save()
+    """Reassigning FSMStates"""
+    tmp_user.fsmstate_set.all().update(user=user)
+    """Reassigning Responses"""
+    tmp_user.response_set.all().update(author=user)
+    """Reassigning StudentErrors"""
+    tmp_user.studenterror_set.all().update(author=user)
+
+
+def social_merge(tmp_user, user):
+    """Merge UserSocialAuth
+
+    Re-assign UserSocialAuth objects to given user.
+    """
+    tmp_user.social_auth.all().update(user=user)
+    tmp_user.lti_auth.all().update(django_user=user)
+
+
+@partial
+def validated_user_details(strategy, backend, details, user=None, is_new=False, *args, **kwargs):
+    """Merge actions
+
+    Make different merge actions based on user type.
+    """
+    social = kwargs.get('social')
+    email = details.get('email')
+    if user and user.groups.filter(name='Temporary').exists():
+        if social:
+            tmp_user = user
```

Comment:
This can be done w/o new variable. Instead of user=social.user, we can write social_user = social_user.